Adding new Feature/timestep

drnimbusrain opened the pull request:
This PR addresses adding the capability for canopy-app to read/process/write multiple time steps, rather than only instantaneous single hour.

This adds multiple input NC and TXT files to the input/ directory for examples in the SE domain, where by default the output is put into single NC file with multiple time steps for NC file output.  For TXT output, canopy-app sends the output to multiple txt files with included valid timestamp on the filename.

Here is an example for input downward shortwave flux and 2-m temperature, and output summed ISOP emissions valid @ July 01, 2022 at 11, 12, and 13 UTC:

**Input Downward Shortwave Flux:**
July 01 @ 11 UTC
![image](https://github.com/noaa-oar-arl/canopy-app/assets/26631222/c8a88055-5e25-4a81-9add-bee5f13f3428)
July 01 @ 12 UTC
![image](https://github.com/noaa-oar-arl/canopy-app/assets/26631222/ff49f335-bfde-40df-8dc1-8939c03ebaf4)
July 01 @ 13 UTC
![image](https://github.com/noaa-oar-arl/canopy-app/assets/26631222/17e17823-7a31-4f7b-9b68-375a4da58701)
**Input 2-m Temperature**
July 01 @ 11 UTC
![image](https://github.com/noaa-oar-arl/canopy-app/assets/26631222/3e5460fe-f48b-40ff-8f31-6bf05573af52)
July 01 @ 12 UTC
![image](https://github.com/noaa-oar-arl/canopy-app/assets/26631222/4b218316-05d1-4979-b478-55e516c8d873)
July 01 @ 13 UTC
![image](https://github.com/noaa-oar-arl/canopy-app/assets/26631222/8fab934e-31f7-46c9-b4a0-57fff4389355)
**Output Summed ISOP Biogenic Emissions**
July 01 @ 11 UTC
![image](https://github.com/noaa-oar-arl/canopy-app/assets/26631222/96f35614-356a-4d4c-9140-d27c8d1b0b12)
July 01 @ 12 UTC
![image](https://github.com/noaa-oar-arl/canopy-app/assets/26631222/15466a17-6f7d-4a07-abf0-c8c67b8744c2)
July 01 @ 13 UTC
![image](https://github.com/noaa-oar-arl/canopy-app/assets/26631222/b1062688-3897-4d45-bbde-fefc2bc4afd2)

@zmoon Since the new SE example files cover three time periods, I assume this will break the current CI?

drnimbusrain committed:
Initial commit of feature/timestep branch.

drnimbusrain committed:
Merge branch 'develop' into feature/timestep

drnimbusrain committed:
Merge branch 'develop' into feature/timestep

drnimbusrain committed:
Updated time step feature and new canopy_close_files.

drnimbusrain committed:
Changes to canopy_app structure and fixed firsttime issue.

drnimbusrain committed:
Added time attributes to output file.

drnimbusrain committed:
Updated multiple timesteps, remove closefiles, other changes.

drnimbusrain committed:
Added forgotten date_mod codes.

drnimbusrain committed:
Added timestamps to nc and txt file outputs.

drnimbusrain committed:
Added point example to namelist for completeness.

drnimbusrain commented:
@zmoon Know your super busy, but maybe start a review of this next week ?

Thank you!

drnimbusrain committed:
Reverting default NL to single file for CI test.

drnimbusrain committed:
Updated python script due to txt output file name changes.

drnimbusrain committed:
Added new time stamp suffix to txt output for python script.

drnimbusrain commented:
Starting playing with changes to NL for CI and the python/canopy_app.py script.  However, still brakes the python testing as it doesn't like the new time suffixes placed on the end of the txt files it seems. @zmoon Any suggestions?

quaz115 commented:
> Starting playing with changes to NL for CI and the python/canopy_app.py script. However, still brakes the python testing as it doesn't like the new time suffixes placed on the end of the txt files it seems. @zmoon Any suggestions?

Are the latest changes for passing CI/CD tests?

drnimbusrain commented:
> > Starting playing with changes to NL for CI and the python/canopy_app.py script. However, still brakes the python testing as it doesn't like the new time suffixes placed on the end of the txt files it seems. @zmoon Any suggestions?
> 
> Are the latest changes for passing CI/CD tests?

@quaz115 Yes, but still unsuccessful, would like @zmoon help here as he put the tests in originally.  Its something in the python/canopy_app.py script with my new time stamp suffixes put on the output files  here.  Zach please make any necessary adjustments to CI python scripts directly to make this work.  Thank you!

zmoon committed:
Timed text files

but not parsing the time from the file name yet

zmoon committed:
Timed nc files

zmoon commented:
@drnimbusrain I fixed the output file recognition, but haven't dealt with possible multiple files/times yet.

I am not a huge fan of having the whole time step in the file name. And `:` aren't allowed in paths on Windows which could be a pain. Two possible solutions:
1. Simplified time stamp, e.g. `2022-07-01-12:00:00.0000` -> `2022-07-01_12` (or `2022-07-01_12-00` or such if need to have minute capability)
2. Have the automatic part just be integer step relative to run start (`t000`, `t001`, etc.), while user can still control the prefix with `file_out` nl setting. The time is already in the nc files. For the text files, could add a line like the reference height one.

drnimbusrain committed:
Revised time step/stamp outputs for TXT and NC.

drnimbusrain commented:
> @drnimbusrain I fixed the output file recognition, but haven't dealt with possible multiple files/times yet.
> 
> I am not a huge fan of having the whole time step in the file name. And `:` aren't allowed in paths on Windows which could be a pain. Two possible solutions:
> 
> 1. Simplified time stamp, e.g. `2022-07-01-12:00:00.0000` -> `2022-07-01_12` (or `2022-07-01_12-00` or such if need to have minute capability)
> 2. Have the automatic part just be integer step relative to run start (`t000`, `t001`, etc.), while user can still control the prefix with `file_out` nl setting. The time is already in the nc files. For the text files, could add a line like the reference height one.

@zmoon Given you good suggestions, I decided to go with your option 2 for the reasons you described.  Hopefully we can move this PR forward.  Thanks again!

drnimbusrain committed:
Updated README

drnimbusrain committed:
Updated README

drnimbusrain committed:
Fixed write_txt error for missing argument.

drnimbusrain committed:
Changed python/canopy_app.py for new text headers.

drnimbusrain committed:
Trying this way.

drnimbusrain committed:
Updated skiprows.

zmoon commented on src/canopy_ncf_io_mod.F90:
```diff
@@ -347,7 +347,7 @@ SUBROUTINE canopy_outncf_init
         g_time%cartesian_axis = "T"
         g_time%long_name = 'time'
-        g_time%units = 'hours since ' // time_start
+        g_time%units = 'time steps since ' // time_start
```
Can we have this in seconds or hours so that it is CF compliant? maybe integer steps array multiplied by the step size

drnimbusrain replied:
Changed.

drnimbusrain committed:
Changed time units to "seconds since"

zmoon committed:
Read time stamp from text files

zmoon committed:
Fix nc reading for now

zmoon committed:
Merge branch 'feature/timestep' of https://github.com/noaa-oar-arl/canopy-app into feature/timestep

zmoon committed:
Back to decoding nc times

drnimbusrain committed:
Fixed output time.

drnimbusrain committed:
Merge branch 'feature/timestep' of https://github.com/noaa-oar-arl/canopy-app into feature/timestep

drnimbusrain commented:
@zmoon   Can We move this forward, would like to get this capability into canopy-app soon for testing other updates.  Thanks!

zmoon commented:
@drnimbusrain how should I modify the default namelist to test the multiple timestep txt output?

drnimbusrain commented:
> @drnimbusrain how should I modify the default namelist to test the multiple timestep txt output?

@zmoon Maybe I misunderstand your question, but you can look at how to simply run the default three timesteps for NC or TXT file in the branches updated namelist:   https://github.com/noaa-oar-arl/canopy-app/blob/feature/timestep/input/namelist.canopy

zmoon commented:
Like this commented line (seems like the default is still one time step)?

https://github.com/noaa-oar-arl/canopy-app/blob/33e90682cea11c089608637b22e4992668799586/input/namelist.canopy#L7

Was kind of hoping for point example.

drnimbusrain commented:
> Like this commented line (seems like the default is still one time step)?
> 
> https://github.com/noaa-oar-arl/canopy-app/blob/33e90682cea11c089608637b22e4992668799586/input/namelist.canopy#L7
> 
> Was kind of hoping for point example.

Oh, I just commented that out and reverted to single timestep to see if it would help the current CI pass for now.  The line commented out is indeed the example for hopefully the new default three timestep example:

```
file_vars    = 'input/gfs.t12z.20220630.sfcf023.canopy.txt' 'input/gfs.t12z.20220701.sfcf000.canopy.txt' 'input/gfs.t12z.20220701.sfcf001.canopy.txt'
```

With other related NL text settings for the three times to match:
```
  infmt_opt   =  1
  time_start  = '2022-07-01-11:00:00.0000'
  time_end    = '2022-07-01-13:00:00.0000'
  ntime       =  3
  time_intvl  =  3600
```

Should have explanations of these new NL time variables in the [README](https://github.com/noaa-oar-arl/canopy-app/blob/feature/timestep/README.md).

drnimbusrain committed:
Changed NL back to 3 times and added point example.

drnimbusrain commented:
@zmoon   But I see now what you are asking.   I reverted back to the three times as default NL for your CI updates here, and provide an example point file for three times as well in latest [commit](https://github.com/noaa-oar-arl/canopy-app/pull/87/commits/ce6db12a84eda8990afbf12175616d6306e676fa).    Let me know if this works for your changes/testing.  Note I change the example point input file names as well.

zmoon committed:
Keep default case in CI as single file

zmoon committed:
Fix setup for multiple input files

zmoon committed:
Deal with multiple times in output txt files

zmoon committed:
ntime

drnimbusrain committed:
Updated point files to change vtype and fix fch.

drnimbusrain committed:
Merge branch 'feature/timestep' of https://github.com/noaa-oar-arl/canopy-app into feature/timestep

zmoon committed:
Replace fill value with NaN when reading txt outputs

zmoon committed:
Fix examples

- need to specify ntime now
- different canopy height for point
- for cases generation, limit to single time step for now
  - since file_vars can be str OR list(str) now, hard to know
    if it is list of options or a single config

only one time since CI currently overriding default nl

recall squeezed before returning

drnimbusrain committed:
Updated SE txt and nc files for LAI updates from Wei-Ting.

drnimbusrain committed:
Removed timestamp from updated default SE txt files.

drnimbusrain commented:
@zmoon 

@quaz115  is wondering how this PR is going, so he can merge and test with the updated biogenics and leafage, with some aspect of saving previous times (for LAI inputs).

zmoon reviewed (approved):
@drnimbusrain I added a point example with multiple output times. I think it is fine but feel free to check.

I hope to simply the Python code for dealing with the multiple times (set `file_vars` as length-1 list instead of string when single file) another time.
